datafusion/physical-expr/src/window/rank.rs

```diff
@@ -98,18 +99,77 @@ impl BuiltInWindowFunctionExpr for Rank {
         &self.name
     }
 
+    fn supports_bounded_execution(&self) -> bool {
+        matches!(self.rank_type, RankType::Basic | RankType::Dense)
+    }
+
     fn create_evaluator(&self) -> Result<Box<dyn PartitionEvaluator>> {
         Ok(Box::new(RankEvaluator {
+            state: RankState::default(),
             rank_type: self.rank_type,
         }))
     }
 }
 
+#[derive(Debug)]
 pub(crate) struct RankEvaluator {
+    state: RankState,
     rank_type: RankType,
 }
 
 impl PartitionEvaluator for RankEvaluator {
+    fn get_range(&self, state: &WindowAggState, _n_rows: usize) -> Result<Range<usize>> {
+        Ok(Range {
+            start: state.last_calculated_index,
+            end: state.last_calculated_index + 1,
+        })
+    }
+
+    fn state(&self) -> Result<BuiltinWindowState> {
+        Ok(BuiltinWindowState::Rank(self.state.clone()))
+    }
+
+    fn update_state(
```

Comment:
While reading through these implementaitons I can't help but wonder if there is some way to avoid implementing each PartitionEvaluator multiple times (`evaluate_stateful` seems like a more generic version of `evaluate` where the state is explicitly tracked rather than in the local variables).

I can't remember why we also needed `evaluate_with_rank`

Maybe the window functions can eventually look like the aggregators where they all have some state and there are functions to update that state / merge that state. 🤔 

It appears you somewhat did this in `datafusion/physical-expr/src/window/sliding_aggregate.rs` where the accumulator is used in both `evaluate` and `evaluate_stateful`

Reply:
I agree that, we can make the implementation of Builtin WindowFunctions more similar to the aggregators. This will possibly increase code re-usage. We will consider this point in follow-up PRs.